db/nft.go

```diff
@@ -381,135 +381,91 @@ func GetNftEvents(conn *pgxpool.Conn, q QueryEventsRequest, p PageRequest) (Quer
+		var i NftIncomeResponse
+
+		if err = rows.Scan(&ci.ClassId, &ci.CreatedAt, &i.Address, &i.Amount, &ci.Sales, &res.Pagination.NextKey); err != nil {
 			logger.L.Errorw("failed to scan nft incomes", "error", err, "q", q)
 			return QueryIncomesResponse{}, fmt.Errorf("query nft incomes data failed: %w", err)
 		}
-		res.Incomes = append(res.Incomes, r)
-		res.TotalAmount += r.Amount
-	}
-	res.Pagination.Count = len(res.Incomes)
-	return res, nil
-}
-
-func GetNftIncomeDetails(conn *pgxpool.Conn, q QueryIncomeDetailsRequest, p PageRequest) (QueryIncomeDetailsResponse, error) {
-	ownerVariations := utils.ConvertAddressPrefixes(q.Owner, AddressPrefixes)
-	stakeholderVariations := utils.ConvertAddressPrefixes(q.Address, AddressPrefixes)
-	orderBy := "i.id"
-	switch q.OrderBy {
-	case "price":
-		orderBy = "e.price"
-	case "income":
-		orderBy = "i.amount"
-	case "default":
-	default:
-		orderBy = "i.id"
-	}
 
-	sql := fmt.Sprintf(`
-		SELECT e.class_id, e.nft_id, e.tx_hash, e.timestamp, i.address, 
-			e.price, i.amount 
-		FROM nft_event AS e
-		JOIN nft_class as c
-			ON e.class_id = c.class_id
-		JOIN iscn
-			ON iscn.iscn_id_prefix = c.parent_iscn_id_prefix
-		JOIN iscn_latest_version
-			ON iscn.iscn_id_prefix = iscn_latest_version.iscn_id_prefix
-			AND iscn.version = iscn_latest_version.latest_version
-		JOIN nft_income AS i
-			ON e.class_id = i.class_id 
-			AND e.nft_id = i.nft_id
-			AND e.tx_hash = i.tx_hash
-		WHERE ($2 = 0 OR i.id > $2)
-			AND ($3 = 0 OR i.id < $3)
-			AND ($4 = '' OR e.class_id = $4)
-			AND ($5 = '' OR e.nft_id = $5)
-			AND ($6::text[] IS NULL OR cardinality($6::text[]) = 0 OR iscn.owner = ANY($6))
-			AND ($7::text[] IS NULL OR cardinality($7::text[]) = 0 OR i.address = ANY($7))
-			AND ($8 = 0 OR (e.timestamp IS NOT NULL AND e.timestamp > to_timestamp($8)))
-			AND ($9 = 0 OR (e.timestamp IS NOT NULL AND e.timestamp < to_timestamp($9)))
-			AND ($10::text[] IS NULL OR cardinality($10::text[]) = 0 OR e.action = ANY($10))
-		ORDER BY %[1]s %[2]s
-		LIMIT $1
-	`, orderBy, p.Order())
-
-	ctx, cancel := GetTimeoutContext()
-	defer cancel()
-
-	rows, err := conn.Query(
-		ctx, sql,
-		p.Limit, p.After(), p.Before(), q.ClassId, q.NftId,
-		ownerVariations, stakeholderVariations, q.After, q.Before, q.ActionType,
-	)
-	if err != nil {
-		logger.L.Errorw("Failed to query nft income details", "error", err)
-		return QueryIncomeDetailsResponse{}, fmt.Errorf("query nft income details error: %w", err)
-	}
-
-	res := QueryIncomeDetailsResponse{
-		IncomeDetails: make([]NftIncomeDetailResponse, 0),
-	}
-	for rows.Next() {
-		var r NftIncomeDetailResponse
-		if err = rows.Scan(
-			&r.ClassId, &r.NftId, &r.TxHash, &r.Timestamp, &r.Address,
-			&r.Price, &r.Amount,
-		); err != nil {
-			logger.L.Errorw("failed to scan nft income details", "error", err, "q", q)
-			return QueryIncomeDetailsResponse{}, fmt.Errorf("query nft income details data failed: %w", err)
+		if ciPtr == nil || ci.ClassId != ciPtr.ClassId { // is new class
```

Comment:
instead of using pointer, can't we just make a map and assign/append data accordingly?

Reply:
Tried using a `rankMap := make(map[int]NftClassIncomeResponse)` to collect the row data. When converting the map to response data, it seems the iteration order of `range` operation on `map` is not guaranteed in Go. Thus [additional sorting would be required](https://stackoverflow.com/a/18342865/7978205) and make time complexity change from O(N) to O(NlogN)?

Reply:
@nnkken please comment on how to write this best

Reply:
better rename to `ciPrev`

Reply:
using a map will be clearer, here the data size is small (hundreds at most I guess?) so a sorting is acceptable.
We may also have an extra array of map keys, and append to this array only if it does not exist in the map, to maintain insert order.

Reply:
Refactored